src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs

```diff
@@ -374,39 +389,39 @@ public void SendAsyncCancel()
 
         private PingReply GetAddressAndSend(string hostNameOrAddress, int timeout, byte[] buffer, PingOptions options)
         {
+            CheckStart();
             try
             {
                 IPAddress[] addresses = Dns.GetHostAddresses(hostNameOrAddress);
                 return SendPingCore(addresses[0], buffer, timeout, options);
             }
             catch (Exception e)
             {
-                Finish();
                 throw new PingException(SR.net_ping, e);
             }
+            finally
+            {
+                Finish();
+            }
         }
 
         private async Task<PingReply> GetAddressAndSendAsync(string hostNameOrAddress, int timeout, byte[] buffer, PingOptions options)
         {
-            bool requiresFinish = true;
+            CheckStart();
```

Comment:
The caller is calling CheckDisposed, and then CheckStart is calling CheckDisposed again.  Is that on purpose?

Reply:
I removed CheckDisposed from CheckStart and updated rest of the code. Now CheckStart only does state processing balanced out in Finish()